### PR TITLE
Extend functions/macros not allowed inside Protocols

### PR DESCRIPTION
**Pull request by eksperimental**
Add to the list: defexception/1, defdelegate/2, defguard/1, defguardp/1

### CODE REVIEW

File: lib/elixir/test/elixir/protocol_test.exs [hunk 1]
```diff
@@ -262,6 +262,33 @@ defmodule ProtocolTest do
...
+        end
+      end
+    end
+
+    assert_raise CompileError, ~r"undefined function defguard/1", fn ->
+      defprotocol Foo do
+        defguard is_something(arg) when arg == :something
+      end
+    end
+  end
```
Comment (josevalim) [inline lib/elixir/test/elixir/protocol_test.exs:291]:
```suggestion
```

Comment (josevalim) [reply lib/elixir/test/elixir/protocol_test.exs:291]: We don't need those tests as they are basically testing the declarative aspect of the code.

**Comment by eksperimental** [comment]
This is probable for another PR. I was trying to defined a __using__/1 macro iside a protocol, and I was scratching my head becuase it said defmacro/2 was not defined. I had to go to the source code of protocol and figure out what was going on.
I think we should override these functions and emit a warning when called, to give the user a hint of what is going on.
On the other hand, there is no documentation that backs this behaviour.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**